### ISSUE TITLE
Android Compose: placeholder + crossfade = the image isn't cropped bug

### ISSUE THREAD

**Issue by tiagonuneslx** [coil-kt/coil#855]
**Describe the bug**
On Jetpack Compose, when displaying an image with a placeholder and crossfade, the image isn't cropped.

**Expected behavior**
The placeholders should appear clipped until the images are loaded. Then, the images should appear also clipped, after a crossfade effect.

**To Reproduce**
Code to reproduce:
https://gist.github.com/tiagonuneslx/d3eda749c0d4687a04706402755c2545#file-coil-compose-crop-placeholder-crossfade-bug-kt
Simple project reproducing bug: https://github.com/tiagonuneslx/coil-compose-crop-placeholder-crossfade-bug

**Logs/Screenshots**

Without placeholder + crossfade:

![image](https://user-images.githubusercontent.com/49025614/128934754-12773d84-be82-410f-a716-2a0dbfdc1bdf.png)

With placeholder + crossfade:

![image](https://user-images.githubusercontent.com/49025614/128934798-b2070e54-0781-4cf1-b1d7-8bdbb10de1c6.png)

**Version**
io.coil-kt:coil-compose:1.3.2 (Android Jetpack Compose)

**Comment by rsookram** [comment]
**TL;DR** To work around this for now, the size of the placeholder can be reduced to be less than the size of the loaded image. In the sample project, this can be done by reducing [`width` and `height` of `ic_food`](https://github.com/tiagonuneslx/coil-compose-crop-placeholder-crossfade-bug/blob/f91781f988121396808043220ef6a117d8b30987/app/src/main/res/drawable/ic_food.xml#L1-L3).

I tried looking into this, and it seemed to be happening because the placeholder image was bigger (and had a different aspect ratio) than the loaded image. This meant that the intrinsic size of the `CrossfadePainter` was [based off the size of the placeholder](https://github.com/coil-kt/coil/blob/7db786a1ae721bd8a14d52ace4f286335e27bc3c/coil-compose-base/src/main/java/coil/compose/CrossfadePainter.kt#L99-L102).

It was useful to add `elevation` to the `Surface` to see what was actually happening:

<img width=250 src="https://user-images.githubusercontent.com/1985558/130328160-8ec5ef75-8af8-43b8-b1eb-04a28c102ea5.png" alt="Sample image with dropshadow around the Surface which performs rounding">

Upon further debugging, the incorrect cropping was happening because the [image was getting inset](https://github.com/coil-kt/coil/blob/7db786a1ae721bd8a14d52ace4f286335e27bc3c/coil-compose-base/src/main/java/coil/compose/CrossfadePainter.kt#L118-L121) when being drawn. It's not clear to me why this inset is needed though. Maybe it's for different `ContentScale`s?

**Comment by colinrtwhite** [comment]
@rsookram Thanks for taking a look! Do you think there's a way we could better handle this case in `CrossfadePainter`?

**Comment by rsookram** [comment]
@colinrtwhite I read more of the code in `CrossfadePainter`, and I think I have a better understanding of what's happening now. I don't have a good solution though :disappointed:

Previously, I wrongly assumed that [`computeDrawSize`](https://github.com/coil-kt/coil/blob/7db786a1ae721bd8a14d52ace4f286335e27bc3c/coil-compose-base/src/main/java/coil/compose/CrossfadePainter.kt#L127) was using the `contentScale` that was specified on the `Image`, but it turns out that it's using Coil's `Scale` type. This `Scale` is [defaulting to `FIT`](https://github.com/coil-kt/coil/blob/7db786a1ae721bd8a14d52ace4f286335e27bc3c/coil-compose-base/src/main/java/coil/compose/ImagePainter.kt#L353-L354) in this case, because it wasn't specified in the request builder. When the scale is specified as `FILL` in the request builder, the problem is also solved (this is probably better than changing the placeholder size like I mentioned in my previous comment).

It seems like you came across something similar with `ImagePainter`, but that class is [defaulting to `FILL`](https://github.com/coil-kt/coil/blob/7db786a1ae721bd8a14d52ace4f286335e27bc3c/coil-compose-base/src/main/java/coil/compose/ImagePainter.kt#L231-L235) instead. I'm not sure why this case should behave differently however.

Regardless of the default scale type though, it seems like there's a mismatch in the APIs provided by Compose and Coil here. I think that forcing consumers to specify the scale type in the request builder would solve the problem, but it would make for a less ergonomic API. Documenting this behavior would be better than nothing, but I figure that some people would miss it.

It's possible that there's some better way of handling this that I don't know of yet, but otherwise, maybe doing the crossfade in a painter isn't the preferred way to handle this in Compose?

**Comment by colinrtwhite** [comment]
@rsookram Thanks for taking a look! You're right, we can end up with a mismatch between Coil's scale and the scale used by the `Image` composable. This can also technically happen with `ImageView`s, however using `load` or another API lets Coil grab the current scale type and use it for the request. Unfortunately there's no API to get the `ContentScale` for the `Image` composable so `rememberImagePainter` has to default to `Scale.FIT` (analogous to the default `ContentScale`). We default to `Scale.FILL` [here](https://github.com/coil-kt/coil/blob/7db786a1ae721bd8a14d52ace4f286335e27bc3c/coil-compose-base/src/main/java/coil/compose/ImagePainter.kt#L231-L235) to ensure that the image is always greater than or equal to the necessary size to be rendered.

I don't think we can force setting the `scale` in the request builder since it would be a breaking change, but we should probably better document this behaviour as you mentioned. Ideally, we'd be able to get the `ContentScale` inside the painter.

@tiagonuneslx Thanks for the report + repro project. Unfortunately, I don't think there's a way for Coil to automatically fix this without breaking changes. Adding `scale(Scale.FILL)` to your request builder should fix the issue.

**Comment by colinrtwhite** [comment]
Going to close this out. I've added documentation for this to the 2.x branch. Unfortunately I don't think there's a better solution than asking users to specify `scale` if they use a non-default `ContentScale`, though I'm still brainstorming...

**Comment by colinrtwhite** [comment]
@tfcporciuncula Hmm I can't seem to find the docs I was referencing either. Though, I think `AsyncImage` should handle this case properly now - it's not possible to handle this case automatically with `rememberAsyncImagePainter`.

**Comment by tfcporciuncula** [comment]
Thanks for checking! I think I've reached a more specific case where I have scaling issues when I have a placeholder with different dimensions (and the scaling issues only show up with crossfade). I'm filling a new issue [here](https://github.com/coil-kt/coil/issues/1688).
